server/app/watchlist.py

```diff
@@ -56,6 +57,53 @@ def __iter__(self):
 	def is_invalid(self):
 		return self.__is_invalid
 
+	def write(self):
+		self.write_watchlist(self, self.path)
+
+	def add_channel(self, id=None, source='youtube', site=None, name='channel'):
+		if id is None:
+			raise TypeError('Missing keyword argument "id", string expected')
+		source = source.lower()
+		if source not in url_formats:
+			raise ValueError('source not in url formats')
+		if site is None:
+			site = url_formats[source]['default_site']
+		self.channels.append({
+			'name': name,
+			'id': id,
+			'source': source,
+			'site': site,
+		})
+
+	def remove_channel(self, id=None, source=None, site=None, name=None):
+		if id is None:
+			raise TypeError('Missing keyword argument "id", string expected')
+		for i in range(len(self.channels)):
+			channel = self.channels[i]
+			if channel['id'] != id:
+				continue
+			if source is not None and channel['source'] != source.lower():
+				continue
+			if site is not None and channel['site'] != site:
+				continue #pragma nocover
+			if name is not None and channel['name'] != name:
+				continue #pragma nocover
+			self.channels.pop(i)
+			return True
+		return False
+
+	@staticmethod
+	def write_watchlist(watchlist, path):
+		config = ConfigParser()
+		for channel in watchlist.channels:
+			config[channel['name']] = {
```

Comment:
this is one example how using non-unique `name` instead of inherently-unique `id` as a label will cause bugs.

Reply:
oh shoot, I didn't catch that. I thought I was appending.
I'll just change the labels to be in the format of '{name} ({id})'.
Then we can still have the option to put the channel title in the label while having them be unique

Reply:
Eh, that still leaves the door open for weird bugs like having 2 entries for the same channel id but with different names. If you really want the option to see the channel title in the file, let's keep `name` as a property but only use `id` as an indexing label.

Reply:
No, name has to be the indexing label for python's configparser library.
I'm going to have the watchlist handle the name property completely internally,
and not allow any part of the api to specify name. When deleting from the watchlist,
only "id" will be used for matching.

Reply:
Nevermind, I'm not thinking straight. It'll take a little more work, but I can just use 'id' as a label and name as a property.

Reply:
not sure what you mean, but why not do the following:
```
config[channel['id']] = {
    'name': channel['name'],
    'source': channel['source'],
    'site': channel['site'],
}
```

Reply:
Yeah, that's what I meant. I just didn't think of it before.